### PR TITLE
NXP-28918: add a 'nx:arrayContains' function in the EL context

### PR DESCRIPTION
**Pull request by troger**
This function is used to test if an array contains a given element.

### CODE REVIEW

**Comment by troger** [comment]
T&P: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-troger/448/

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Helper functions injected in the {@link ExpressionContext} instance.
+ *
+ * @since 11.2
+ */
+public class Functions {
+
+    /**
+     * Returns {@code true} if the given {@code arr} contains the specified element.
```
Comment (ataillefer) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java:32]: For consistency:
Returns {@code true} if the given {@code arr} contains the given {@code o}.
or
Returns {@code true} if the given array contains the given object.

File: modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java [hunk 2]
```diff
@@ -106,4 +107,21 @@ public void testList() {
         assertNotNull(value);
     }
 
+    /**
+     * NXP-28918
+     */
```
Comment (ataillefer) [inline modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java:112]: Could add `@since 11.2`

File: modules/platform/nuxeo-platform-imaging/nuxeo-platform-imaging-core/src/test/java/org/nuxeo/ecm/platform/picture/core/test/TestPictureConversions.java [hunk 3]
```diff
@@ -194,6 +194,32 @@ public void shouldFilterPictureConversions() throws Exception {
         assertNull(multiviewPicture.getView("anotherSmallConversion"));
     }
 
+    /**
+     * NXP-28918
+     */
```
Comment (ataillefer) [inline modules/platform/nuxeo-platform-imaging/nuxeo-platform-imaging-core/src/test/java/org/nuxeo/ecm/platform/picture/core/test/TestPictureConversions.java:199]: `@since 11.2`

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java (hunk 1 above)
Comment (efge) [reply modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java:32]: You should specify in the Javadoc what happens for a `null` array (it's fine to make it behave like an empty one).

File: modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java [hunk 4]
```diff
@@ -106,4 +107,21 @@ public void testList() {
...
+    @Test
+    public void testArrayContainsFunction() {
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java:114]: Please also test the `null` case.

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    public static boolean arrayContains(Object[] arr, Object o) {
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java:34]: While we're at it wouldn't it be useful to also code now an `arrayContainsAll` and `arrayContainsAny`? And maybe `arrayContainsNone`?

Comment (troger) [reply modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/Functions.java:34]: I'm not always a fan of adding methods that are not needed but could be a good idea here.

File: modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java (hunk 4 above)
Comment (troger) [reply modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java:114]: WDYM? The array to `null` or the element?

Comment (efge) [reply modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java:114]: I meant the array.

Comment (efge) [reply modules/core/nuxeo-core-el/src/test/java/org/nuxeo/platform/el/TestExpressionEvaluator.java:114]: Ah yes sorry

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java [hunk 6]
```diff
@@ -30,6 +32,15 @@
 
 public class ExpressionContext extends ELContext {
 
+    private static final Map<String, Method> DEFAULT_FUNCTIONS = new HashMap<>();
+
+    static {
+        Stream.of(Functions.class.getMethods())
+              .filter(m -> Modifier.isStatic(m.getModifiers()))
+              .filter(m -> Modifier.isPublic(m.getModifiers()))
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java]: Not needed, `getMethods` already returns only `public` methods.

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java [hunk 7]
```diff
@@ -30,6 +32,15 @@
 
 public class ExpressionContext extends ELContext {
 
+    private static final Map<String, Method> DEFAULT_FUNCTIONS = new HashMap<>();
+
+    static {
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java]: You could use `Collectors.toUnmodifiableMap` to directly define the value of `DEFAULT_FUNCTIONS` without a static block.

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java [hunk 8]
```diff
@@ -51,7 +62,13 @@ public ValueExpression setVariable(String variable, ValueExpression expression)
 
         @Override
         public Method resolveFunction(String prefix, String localName) {
-            return map.get(prefix + ":" + localName);
+            String key = String.format("%s:%s", prefix, localName);
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java:65]: `String.format` is overkill and costly (and you don't use it in `mapFunction` anyway).

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java [hunk 9]
```diff
@@ -30,6 +35,12 @@
 
 public class ExpressionContext extends ELContext {
 
+    private static final Map<String, Method> DEFAULT_FUNCTIONS = //
+            Stream.of(Functions.class.getMethods())
+                  .filter(m -> Modifier.isStatic(m.getModifiers()))
+                  .filter(m -> Modifier.isPublic(m.getModifiers()))
```
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java:41]: No need for `isPublic`, as `getMethods` only returns public methods

File: modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java (hunk 8 above)
Comment (efge) [inline modules/core/nuxeo-core-el/src/main/java/org/nuxeo/ecm/platform/el/ExpressionContext.java:65]: `String.format` is overkill and costly (and you don't use it in `mapFunction` anyway).

**Comment by troger** [comment]
T&P following changes: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-troger/449/
